File: src/models/message.py
from datetime import datetime
from typing import Literal

from pydantic import BaseModel, Field, field_validator


class Message(BaseModel):
    id: str
    world_id: str
    type: str  # dialogue / narration / system / user
    sender_type: str  # character / narrator / system / user
    sender_id: str | None = None
    sender_name: str | None = None
    content: str
    real_time: datetime | None = None
    is_key_message: bool = False
    user_participated: bool = False
    created_at: datetime | None = None
    session_id: str | None = None
    sequence: int | None = None
    idempotency_key: str | None = None
    status: str = "normal"

    @field_validator("type")
    @classmethod
    def validate_type(cls, v: str) -> str:
        allowed = {"dialogue", "narration", "system", "user", "event"}
        if v not in allowed:
            raise ValueError(f"type must be one of {allowed}, got '{v}'")
        return v

    @field_validator("sender_type")
    @classmethod
    def validate_sender_type(cls, v: str) -> str:
        allowed = {"character", "narrator", "system", "user"}
        if v not in allowed:
            raise ValueError(f"sender_type must be one of {allowed}, got '{v}'")
        return v


class SendMessageRequest(BaseModel):
    model_config = {"extra": "forbid"}

    content: str = Field(..., min_length=1)
    participant_mode: Literal["auto", "edit", "include"] = "auto"
    participants: list[dict] | None = None
    session_id: str | None = None
    memories_enabled: bool = False
    action_descriptions: bool = False
    show_narration: bool = False  # 是否生成旁白
    element_rerank: bool = False  # AI 元素精排（会话级开关，开启时用副模型打分）
    user_role: str | None = None  # 用户当前扮演的角色 ID
    idempotency_key: str | None = None
    element_injection_ids: list[str] | None = None  # 手动元素注入
    constraint: str | None = None  # 临时约束

    @field_validator("participants")
    @classmethod
    def validate_participants(cls, v: list[dict] | None) -> list[dict] | None:
        """Issue 19: validate participant dicts have required 'id' field."""
        if v is None:
            return v
        for i, p in enumerate(v):
            if not isinstance(p, dict):
                raise ValueError(f"participants[{i}] must be a dict")
            if "id" not in p:
                raise ValueError(f"participants[{i}] is missing required 'id' field")
        return v

    @field_validator("content")
    @classmethod
    def content_not_whitespace(cls, v: str) -> str:
        if not v.strip():
            raise ValueError("content cannot be whitespace only")
        return v


class SendMessageResponse(BaseModel):
    user_message: Message
    responses: list[Message] = []
    narration: Message | None = None
    error: str | None = None
    session_id: str | None = None
    participants: list[dict] = []
    participant_mode: Literal["auto", "edit", "include"] = "auto"
    memory_flush_triggered: bool = False


class MessageListResponse(BaseModel):
    messages: list[Message]
    has_more: bool

File: frontend/src/types/message.ts
export type MessageType = 'dialogue' | 'narration' | 'system' | 'user' | 'event'
export type SenderType = 'character' | 'narrator' | 'system' | 'user'

export interface Message {
  id: string
  world_id: string
  type: MessageType
  sender_type: SenderType
  sender_id: string | null
  sender_name: string | null
  content: string
  real_time: string | null
  is_key_message: boolean
  user_participated: boolean
  created_at: string | null
  sequence?: number | null
  status?: 'normal' | 'failed'
}

// Import into local scope (used below) and re-export for consumers
import type { Participant } from './chatSession'
export type { Participant }

export interface SendMessageRequest {
  content: string
  participant_mode?: 'auto' | 'edit' | 'include'
  participants?: Participant[] | null
  session_id?: string | null
  memories_enabled?: boolean
  action_descriptions?: boolean
  show_narration?: boolean
  element_rerank?: boolean
  idempotency_key?: string | null
  element_injection_ids?: string[] | null
  constraint?: string | null
}

/** Queue item for the message queue system */
export interface QueueItem {
  id: string
  idempotencyKey: string
  content: string
  senderType: 'user'
  participantMode: 'auto' | 'edit' | 'include'
  participants: Participant[] | null
  /** User role character id selected at the moment this item was enqueued (null = explorer) */
  userRole: string | null
  status: 'pending' | 'processing' | 'completed' | 'failed'
  createdAt: number
  retryCount: number
}

/** Shape stored in localStorage for queue persistence */
export interface QueuePersistData {
  worldId: string
  sessionId: string | null
  pendingQueue: QueueItem[]
  processingItem: QueueItem | null
}

export interface SendMessageResponse {
  user_message: Message
  responses: Message[]
  narration: Message | null
  error: string | null
  session_id: string | null
  participants: Participant[]
  participant_mode: 'auto' | 'edit' | 'include'
  memory_flush_triggered?: boolean
}

export interface MessageListResponse {
  messages: Message[]
  has_more: boolean
}

export interface EventCard {
  title: string
  description: string
  participants: string[]
}

export interface SseEventInjectedPayload {
  event_id: string
  card_message_id: string
  title: string
  description: string
  participants: string[]
  card_message_sequence?: number | null
  session_id?: string | null
}

export interface SseSpeakerTurnPayload {
  id: string
  sender_name: string
  sender_id: string | null
  content: string
  sequence?: number | null
}

export interface SseNarratorTurnPayload {
  id: string
  content: string
  sequence?: number | null
}
